recipes/obake/meta.yaml

```diff
@@ -0,0 +1,90 @@
+{% set version = "0.1" %}
+
+package:
+  name: obake
+  version: {{ version }}
+
+source:
+  url: https://github.com/bluescarni/obake/archive/v{{ version }}.tar.gz
+  sha256: 946de817940421d8af8c614c289b74dc6e221ea803fc2b3dacc30e63d7dc3246
+
+build:
+  number: 0
+  skip: true   # [win and vc<14]
+
+requirements:
+  build:
+    - {{ compiler('c') }}
+    - {{ compiler('cxx') }}
+    - cmake
+    - clang      # [win]
+    - ninja      # [win]
+
+outputs:
+  - name: obake
+    build:
+      run_exports:
+        # SO name changes at minor rev bumps.
+        - {{ pin_subpackage('obake', max_pin='x.x') }}
+    requirements:
+      host:
+        - abseil-cpp
+        - tbb-devel
+        - tbb
+        - mppp
+        - boost-cpp
+        - backtrace    # [linux]
+      run:
+        - tbb
+        - {{ pin_compatible("mppp") }}
+        - boost-cpp
```

Comment:
Since the shared lib doesn't need this, you can drop this.

Reply:
You mean the boost dependency? But ``libobake.so`` links to some boost libraries, so if you install the ``obake`` package without the boost libraries being dragged in you won't get a working ``libobake.so``.

Reply:
Does it? From the conda-build output it doesn't appear to be so.

```
   INFO (obake,lib/libobake.so.1.0): Needed DSO lib/libmp++.so.3 found in conda-forge::mppp-0.17-h9603a25_3
   INFO (obake,lib/libobake.so.1.0): Needed DSO lib/libstdc++.so.6 found in defaults::libstdcxx-ng-9.1.0-hdf63c60_0
   INFO (obake,lib/libobake.so.1.0): Needed DSO lib/libbacktrace.so.0 found in conda-forge::backtrace-1.0-hed695b0_1000
   INFO (obake,lib/libobake.so.1.0): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/libc.so.6 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (obake,lib/libobake.so.1.0): Needed DSO lib/libgcc_s.so.1 found in defaults::libgcc-ng-9.1.0-hdf63c60_0
   INFO (obake,lib/libobake.so.1.0): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/librt.so.1 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
   INFO (obake,lib/libobake.so.1.0): Needed DSO x86_64-conda_cos6-linux-gnu/sysroot/lib/ld-linux-x86-64.so.2 found in CDT/compiler package conda-forge::gcc_impl_linux-64-7.3.0-habb00fd_2
```

Reply:
It does on my system (it links to ``libboost_serialization.so``). But I guess conda build does some analysis of the binaries and concludes that no symbols from ``libboost_serialization.so`` are used from ``libobake.so``?

Reply:
What happens in the following scenario:

- package ``foo`` wants to depend on obake and, specifically, it needs to serialize some obake objects. Because of this, ``foo`` will have to link to ``libboost_serialization.so``, even though ``libobake.so`` does not strictly need to link to ``libboost_serialization.so`` because there's no serialization happening within ``libobake.so``;
- the ``foo`` recipe has a host dependency on ``obake-devel``, and a run dependency on ``obake``. Thus, it will build ok but ``foo`` will not work at runtime because there's nothing bringing in ``boost-cpp`` as a runtime dependency.

So this means that I would need to add ``boost-cpp`` as a runtime requirement for ``foo``? Or am I misunderstanding?

Reply:
There's `-Wl,--as-needed` in LDFLAGS in the conda env. Maybe that's the reason.

Reply:
You need to add `boost-cpp` as a `run_exports` of `obake-devel` too. Then both `obake` and `boost-cpp` will be a runtime dependency of `foo`

Reply:
Otherwise, let's say `foo` was built with `obake` build number 0 and `boost-cpp` 1.70. Then `obake` has `boost-cpp` as a transitive dependency and not directly. What happens if `obake` build number 1 uses `boost-cpp` 1.71, but `foo` still wants boost-cpp 1.70

Reply:
That seems to be the reason. But I don't understand the rationale? Just because I happen not to use symbols from a library in ``libobake.so``, it does not mean that using ``obake`` will never require symbols from that library.

For instance, I could have template code that needs symbols from another library which is just not instantiated in ``libobake.so``, but which could be instantiated by some package depending on obake.

Reply:
That's my point though. `foo` has a direct dependency on `boost-cpp` version at the time that `foo` was built. If you rely on transitive `boost-cpp` from `obake`, you will get errors. So, there are two options.

1. Add `boost-cpp` and `obake-devel` in `foo` host.
2. Add `obake-devel` in `foo` host and add `pin_compatible("boost-cpp", max_pin="x.x")` as a `run_exports` in `obake-devel`